quantized tensor: add preliminary support for advanced indexing, try 2

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49346 quantized tensor: add preliminary support for advanced indexing, try 2**

Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

vkuzo committed:
quantized tensor: add preliminary support for advanced indexing, try 2

Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2274,6 +2274,38 @@ def test_empty_batch(self):
+            # single dim, single index
+            x_q_s1 = x_q[:, [0], :, :]
+            x_fp32_s1 = x_fp32[:, [0], :, :]
+            self.assertTrue(torch.allclose(x_fp32_s1, x_q_s1.dequantize()))
```
should we test for equality of x_q_s1 and x_fp32_s1 + quant + dequant with the same qparams? similar for other cases

vkuzo replied:
sorry, what do you mean?  Anything different from `self.assertTrue(torch.allclose(x_fp32_s1, x_q_s1.dequantize()))`?

jerryzh168 replied:
oh I mean:
```
self.assertEqual(x_fp32_s1, x_q_s1.dequantize())
```

or as Raghu suggested, we can compare the quantized tensor
```
x_fp32_s1_q = torch.quantize_per_tensor(x_fp32_s1, ...)
self.assertEqual(x_fp32_s1_q, x_q_s1) # this will compare quantized tensor, including qparams and int_repr
```

raghuramank100 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -290,6 +290,22 @@ Tensor index(const Tensor & self, TensorList indices) {
 }
 
+Tensor quantized_index(const Tensor & self, TensorList indices) {
+  // For now, this is a naive implementation which does dq -> index -> q.
```
We should fail with an error when the tensor is per-channel quantized.

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2274,6 +2274,38 @@ def test_empty_batch(self):
+            # multiple dim, single index
+            x_q_s2 = x_q[:, [0], [2], :]
+            x_fp32_s2 = x_fp32[:, [0], [2], :]
+            self.assertTrue(torch.allclose(x_fp32_s2, x_q_s2.dequantize()))
```
Might it be useful to check that the int_repr, scale and zero-point are identical instead of the fp32 check.

vkuzo replied:
sure, added an additional check for this for the simple case

vkuzo committed:
assert out per-channel on "quantized tensor: add preliminary support for advanced indexing, try 2"


Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

[ghstack-poisoned]

vkuzo committed:
fix nit on "quantized tensor: add preliminary support for advanced indexing, try 2"


Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2274,6 +2274,39 @@ def test_empty_batch(self):
+            x_q_s1 = x_q[:, [0], :, :]
+            x_fp32_s1 = x_fp32[:, [0], :, :]
+            self.assertTrue(torch.allclose(x_fp32_s1, x_q_s1.dequantize()))
+            self.assertTrue(x_q.dtype == x_q_s1.dtype)
+            self.assertTrue(x_q.q_scale() == x_q_s1.q_scale())
+            self.assertTrue(x_q.q_zero_point() == x_q_s1.q_zero_point())
+            self.assertEqual(x_q.int_repr()[:, [0], :, :], x_q_s1.int_repr())
```
we can just quantize x_fp32_s1 and compare the quantized tensor

vkuzo committed:
improve test on "quantized tensor: add preliminary support for advanced indexing, try 2"


Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

[ghstack-poisoned]

vkuzo committed:
rebase on "quantized tensor: add preliminary support for advanced indexing, try 2"


Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -290,6 +290,27 @@ Tensor index(const Tensor & self, TensorList indices) {
+  TORCH_INTERNAL_ASSERT(
+      self.qscheme() == c10::kPerTensorAffine ||
+      self.qscheme() == c10::kPerTensorSymmetric,
+      "Indexing for per-channel quantized Tensors is not supported.");
```
nit: message should probably be only per-tensor quantized Tensor is supported

jerryzh168 reviewed (approved):
LG, thanks for addressing the comments. had one nit comment inline.

vkuzo committed:
Update on "quantized tensor: add preliminary support for advanced indexing, try 2"


Summary:

This is less ambitious redo of
https://github.com/pytorch/pytorch/pull/49129/.

We make the

```
xq_slice = xq[:, [0], :, :]
```

indexing syntax work if `xq` is a quantized Tensor.  For now, we are
making the code not crash, with an in efficient `dq -> index -> q`
implementation.  A future PR can optimize performance by removing
the unnecessary memory copies (which will require some non-trivial
changes to TensorIterator).

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25539365](https://our.internmc.facebook.com/intern/diff/D25539365)

[ghstack-poisoned]
